### ISSUE TITLE
[EDI] Huge memory leak when parsing ~20MB EDI file

### ISSUE THREAD

**Issue by jurabek** [jf-tech/omniparser#157]
Hi, we found a huge memory leak on parsing the EDI file with a size of over 20Mb
- EDI is a sample `INVOIC` type with millions of `LIN` items.
- attaching [SVG of export](https://user-images.githubusercontent.com/8397056/171855465-b7420013-180d-4b7b-b826-2e79c6e2811c.svg) `--memprofile` of benchmark, 
- it allocates over 5GB memory for the single ~23MB EDI file

is there any alternative to rather than calling `ingester.Parse()` for the large files?

**Comment by jf-tech** [comment]
Thx for reporting the issue! A few things:

1. Unfortunately with other things at hand and holiday/vacation, will only get to sit down looking at it in depth in a couple days. Heads up.

2. It would be great we can have your schema and a sample EDI file that repros the problem. If you can zip them and send to  `jf.tech.llc` at gmail. Sanitize if any sensitive info is present.

3. Based on your description, a wild guess is there are millions of `LIN` segments underneath a segment you designate as `is_target`, thus omniparser has no choice but load/keep everything underneath that target segment in memory in case any xpath based queries might hit any of them. "Millions" of segments, and each might take a couple kB can easily lead to GBs of memory. Don't know what solution would be exactly: our current model of parsing EDI is assuming everything under `is_target` is potentially needed thus all in memory.

4. Meanwhile before we can dig deeper, any chance you don't need the full omniparser capability and just a simple EDI segment reader would suffice: https://github.com/jf-tech/omniparser/blob/master/doc/programmability.md#non-validating-edi-segment-reader ?

**Comment by jf-tech** [comment]
@jurabek we haven't received any repro info from you. Still want us to follow up on this?

**Comment by jurabek** [comment]
Hi @jf-tech, it seems `non-validating-edi-segment` could work for us in case of parsing big EDI files, however, I was struggled to parse our line items with split segments into our current transformed Messages 
just wondered, do you have some examples of parsing the whole LIN item using `seg, err := r.Read()` which contains multiple lines?

**Comment by jf-tech** [comment]
Again if you can attach/link/email an small sample that containing the `LIN` segments (with line breakers) so we can take a look.

**Comment by jf-tech** [comment]
Closed due to lack of activities. FYI @jurabek

**Comment by jurabek** [comment]
Hi @jf-tech I emailed for the following email test invoice file

**Comment by jf-tech** [comment]
Opened given new activity.

**Comment by jf-tech** [comment]
A bunch of observations/experiments I did given your sample EDI and schema:

1. Below is the count of all the segments in the sample EDI:
    ```
    $ cat ~/Downloads/omniparser-157/INVOIC_test.EDI | cut -c 1-3 | sort | uniq -c
    1000 BGM
    1000 CUX
    3000 DTM
    36832 FTX
    173522 IMD
    83639 LIN
    48012 MOA
    3041 NAD
    44472 PIA
    44472 PRI
    83639 QTY
    85344 RFF
    47006 TAX
    1 UNA
    1 UNB
    1000 UNH
    1000 UNS
    1000 UNT
    1 UNZ
    ```
    If my tally is correct, in the 16MB sample EDI, we have totally 650K+ segments, all of which are required (based on the
    the instructions of your schema) to be loaded in memory at the same time. Note most of these segments do have
    elements, on avg 3, let's say. In [EDI IDR](https://github.com/jf-tech/omniparser/blob/master/doc/idr.md#edi) each
    segment is constructed as an IDR node, each of its elements is also constructed as an IDR node, then the element
    value another node. So all in all, we are (very roughly) constructing
    `650K * (1 /*seg itself*/ + 3 /*elements*/ + 3 /*element values*/) = 4,550K` IDR nodes.

2. A quick run of of the code in cli mode to do the transform and tallied its consumption:
    ````
    /usr/bin/time -l ./op transform -i ~/Downloads/omniparser-157/INVOIC_test.EDI -s ~Downloads/omniparser-157/invoice.json
            6.11 real        10.92 user         0.33 sys
              2077278208  maximum resident set size
                       0  average shared memory size
                       0  average unshared data size
                       0  average unshared stack size
                  127697  page reclaims
                       0  page faults
                       0  swaps
                       0  block input operations
                       0  block output operations
                       0  messages sent
                       0  messages received
                     675  signals received
                     136  voluntary context switches
                    5797  involuntary context switches
            104770600238  instructions retired
             32099762129  cycles elapsed
              2073549168  peak memory footprint          <<<<<<<<
    ````
    So about 2GB consumed. So that's roughly (again, very roughly) `2G / 4,550K = 468 bytes` for each node. Considering
    the overhead of pointers, other stuff, plus the transformed in-memory `map[interface{}]interface{}` and resulting json
    take memory as well, the each node consumption isn't too outrageous.

3. I inspected the `invoice.json`, some of its elements can be removed since they're not used:
    ```
    $ diff ~/Downloads/omniparser-157/invoice.json ~/Downloads/invoice.json 
    62a63,67
    >                                         "name": "freetext_function",
    >                                         "index": 2,
    >                                         "default": ""
    >                                     },
    >                                     {
    67a73,84
    >                                         "name": "freetext_reference_code",
    >                                         "index": 3,
    >                                         "component_index": 2,
    >                                         "default": ""
    >                                     },
    >                                     {
    >                                         "name": "freetext_reference_code_two",
    >                                         "index": 3,
    >                                         "component_index": 2,
    >                                         "default": ""
    >                                     },
    >                                     {
    134a152,186
    >                                             },
    >                                             {
    >                                                 "name": "party_name",
    >                                                 "index": 3,
    >                                                 "default": ""
    >                                             },
    >                                             {
    >                                                 "name": "party_name",
    >                                                 "index": 3,
    >                                                 "component_index": 1,
    >                                                 "default": ""
    >                                             },
    >                                             {
    >                                                 "name": "party_name",
    >                                                 "index": 3,
    >                                                 "component_index": 2,
    >                                                 "default": ""
    >                                             },
    >                                             {
    >                                                 "name": "party_name",
    >                                                 "index": 3,
    >                                                 "component_index": 3,
    >                                                 "default": ""
    >                                             },
    >                                             {
    >                                                 "name": "party_name",
    >                                                 "index": 3,
    >                                                 "component_index": 4,
    >                                                 "default": ""
    >                                             },
    >                                             {
    >                                                 "name": "party_name",
    >                                                 "index": 3,
    >                                                 "component_index": 5,
    >                                                 "default": ""
    229c281,296
    <                                         "max": -1
    ---
    >                                         "max": -1,
    >                                         "elements": [
    >                                             {
    >                                                 "name": "line_item_identifier",
    >                                                 "index": 1
    >                                             },
    >                                             {
    >                                                 "name": "line_item_number",
    >                                                 "index": 3
    >                                             },
    >                                             {
    >                                                 "name": "line_item_type",
    >                                                 "index": 3,
    >                                                 "component_index": 2
    >                                             }
    >                                         ]
    235a303,307
    >                                                 "name": "product_identifier",
    >                                                 "index": 1,
    >                                                 "default": ""
    >                                             },
    >                                             {
    ```
    That reduced a little memory consumption but not much.

4. `LIN` is indeed not used during transformation. However omniparser EDI parser does require full segment specification in order to validate all the min/max integrity thus it (`LIN`) currently must be included in the schema. I did a quick hack in omniparser EDI reader.go  to exclude `LIN` from being included into the full IDR tree (since it is not used). The resulting `time -l` gives a sizable reduction to 1904711608 total memory used, about 168MB reduction or 8%. Not good enough but it's something. I'll file a separate ticket to add a new attribute `not_in_idr` (or `not_used`, `ignored`, or similar names) for the EDI sgement decl. Default to false, but if it's true, the segment, while must be declared in order for validation to take place, will not be included into the IDR tree (neither will any of its child segments) to reduce memory footprint.

5. That brings me to the final point: while it is possible to redesign the IDR format to compact/compress things up a little bit (the current design of having one new IDR node for each element and one for each of the element values is to make xpath query feasible), but all in all, your situation is: you attempted to load a huge and highly complex hierarchical document in its entirety into memory. omniparser is designed to do parsing/transforming in a "streaming" fashion. While I don't understand fully your business requirements, I couldn't help but noticed: **Instead of putting `is_target=true` on the root segment `UNA`, why not putting it on `UNA/GRP0`, which instantaneously reduces the max memory consumption by more than 1000 folds (since we have about 1000 instances of `UNA/GRP0`) to about 136MB**. You can easily re-combined all those 1000 json blobs back to an array on your side of the code if you'd like to.

Thoughts?

**Comment by jf-tech** [comment]
@jurabek any more investigation/help you need at this point? Or I can close the ticket?

**Comment by jf-tech** [comment]
Closed due to lack of response.
